docs/389ds/design/worker-threads.md

```diff
@@ -0,0 +1,345 @@
+
+### Overload warning (flow control)
+
+A warning is logged if
+
+- The warning has not been logged since some delay (1 minute)
+- The waiting job queue size gets higher than a threshold (100*number of worker threads)
+
+### Producters/consumers algorithm
+
+There is still a global mutex and also three lists:
+
+- waiting_threads
+- busy_threads
+- waiting_jobs
+  Access to these lists are protected by the global mutex lock
+
+When pushing new job the listener threads:
+
+- Get global mutex
+
+- Loop forever:
+  
+  - if waiting_threads is not empty:
+    
+    - Select the first waiting thread
+    - Move the thread out of waiting_threads list into the busy_threads list
+    - Release global mutex
+    - Get thread mutex
+    - Assign job to the thread
+    - Wake up that threads
+    - Release Thread mutex
+    - Abort the loop
+  
+  - else if jobs_free_list is not empty:
+    
+    - Pick an waiting_job list element out of the free list
+    - Set the job in the element
+    - push the element at the end of waiting_job list
+    - Release global mutex
+    - Abort the loop
+  
+  - else:
+    
+    - Release global mutex
+    
+    - if current time > last warning time + warning_delay
+      
+      - Log an user friendly warning about worker threads exhausted and waiting jobs queue full.  (maybe: ***Warning: server may be unresponsive because the threads are exhausted and too many operations have been queued.***)
+      
+      - set last warning time to current time
+    
+    - Sleep 20 ms
+    
+    - Get global mutex
+
+The worker threads while waiting on new work:
+
+- Get the thread mutex
+- Loop forever:
+  - if conn is provided in the thread context
+    - Fill the pblock with conn, op
+    - Reset conn in thread context
+    - Release the thread mutex
+    - return CONN_FOUND_WORK_TO_DO
+  - if thread shutdown flag is set:
+    - Release the thread mutex
+    - return CONN_SHUTDOWN
+  - Release the thread mutex
+  - Get the global mutex
+  - If waiting_job queue is empty
+    - Unlink thread from busy threads list
+    - Link thread at start of waiting threads list
+    - Release the global mutex
+    - Get the thread mutex
+    - Wait on the thread condition variable
+    - Loop again
+      - Move first job from waiting_job list to the jobs_free_list list
+        (after storing the conn in a local variable)
+      - Release the global mutex
+      - Fill the pblock with conn, op
+      - return CONN_FOUND_WORK_TO_DO
+
+### Dynamic change of the number of threads
+
+The number of threads may be changed dynamically by changing the
+ nsslapd-threadnumber attribute in cn=config (when calling config_set_threadnumber). 
+Restart is no more needed.
+
+When changing the number threads:
+
+- if adding threads:
+  - Get global mutex
+  - Compute first new thread index:
```

Comment:
Should check that the server is not in shutdown

Reply:
Fixed (cf lines 155-157)